### PR TITLE
Introduce google_firestore_field resource.

### PR DESCRIPTION
**Pull request by pcostell**
Introduces a single resource for managing Firestore fields, matching the REST API.

Introduces 2 primary divergences from the proto/REST API: 

1) Firestore fields do not support deletion (this is because all fields
   implicitly can exist in a schemaless system). Deletion of a terraform
   resource will clear any "overrides" for this field (TTL or indexing).
2) Single field indexes in the proto API were overly nested to allow
   reusing the same resource for composite indexes. The terraform
   resource unnests the 1-length field in the indexes list.

fixes https://github.com/hashicorp/terraform-provider-google/issues/12151
fixes https://github.com/hashicorp/terraform-provider-google/issues/11419
fixes https://github.com/hashicorp/terraform-provider-google/issues/7593

Tested with ` make testacc TEST=./google TESTARGS='-run=TestAccFirestoreField'`

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_firestore_field`
```

### CODE REVIEW

**Comment by pcostell** [comment]
@rwhogg or @clement would appreciate a Firestore-specific review as well.

File: mmv1/third_party/terraform/tests/resource_firestore_field_test.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/tests/resource_firestore_field_test.go:131]:
```suggestion
    indexes {
        array_config = "CONTAINS"
    }
    indexes {
        order = "DESCENDING"
        query_scope = "COLLECTION_GROUP"
    }
```

A very minor change, but not a blocker to merging as the downstream Go code is formatted during its generation

File: mmv1/products/firestore/Field.yaml
Comment (SarahFrench) [inline mmv1/products/firestore/Field.yaml:81]: I was going to ask about this default value, and whether there should be an acceptance tests that sets `database` to another value, but then I saw this [in the docs](https://cloud.google.com/firestore/docs/reference/rest/v1/projects.databases#resource:-database):

>A Cloud Firestore Database. Currently only one database is allowed per cloud project; this database must have a databaseId of '(default)'.

Noting this on the PR partly for myself as I'm unfamiliar with Firestore, partly for the benefit of anyone looking at the PR in future!

**Review by SarahFrench — APPROVED** [review]
This PR looks good to me in terms of how the downstream code generation works, how the tests are comprehensive, and that the tests pass.

I'm not super familiar with Firestore (I reviewed while referring to the docs) so I'm glad you've tagged others for a Firestore-specific review. Once you've responded to any feedback from them and are satisfied give me an @-mention and I can merge the PR :shipit:

File: mmv1/products/firestore/Field.yaml
Comment (rwhogg) [inline mmv1/products/firestore/Field.yaml:111]: Did you mean to write more here? Looks like this was cut off accidentally.

File: mmv1/products/firestore/Field.yaml
Comment (rwhogg) [inline mmv1/products/firestore/Field.yaml:128]: Should `!=` also be there?

File: mmv1/templates/terraform/custom_check_destroy/firestore_field.go.erb
Comment (rwhogg) [inline mmv1/templates/terraform/custom_check_destroy/firestore_field.go.erb:2]: Is "package custom_check_destroy" included accidentally?

File: mmv1/templates/terraform/custom_flatten/firestore_field_index_config.go.erb
Comment (rwhogg) [inline mmv1/templates/terraform/custom_flatten/firestore_field_index_config.go.erb:18]: Nit: a -> an

File: mmv1/templates/terraform/custom_import/firestore_field.go.erb
Comment (rwhogg) [inline mmv1/templates/terraform/custom_import/firestore_field.go.erb:12]: Nit: please add a space after the `0`

File: mmv1/templates/terraform/examples/firestore_field_basic.tf.erb
Comment (rwhogg) [inline mmv1/templates/terraform/examples/firestore_field_basic.tf.erb:6]: 👍🏻 on all of the examples included. If possible, could we also add one that explicitly shows how to create an SFI exclusion?

**Review by rwhogg — CHANGES_REQUESTED** [review]
Nice work!!! Left some largely-nitpicky comments.

File: mmv1/templates/terraform/examples/firestore_field_basic.tf.erb
Comment (pcostell) [reply mmv1/templates/terraform/examples/firestore_field_basic.tf.erb:6]: Note these are all technically exclusions (they do not inherit from the ancestor configuration). The timestamp example covers the case where a user wants to completely disable their single field indexes. I added a comment there (and updated the description in the yaml) to make this more obvious.

File: mmv1/third_party/terraform/tests/resource_firestore_field_test.go
Comment (pcostell) [reply mmv1/third_party/terraform/tests/resource_firestore_field_test.go]: To clarify, these should be tabs? I made that change (but am having a hard time knowing if it was the right thing to do as it's hard to see the difference in the comment).

**Comment by pcostell** [comment]
Thank you both for the review! Comments should be addressed, PTAL

File: mmv1/products/firestore/Field.yaml
Comment (pcostell) [reply mmv1/products/firestore/Field.yaml:81]: 👍  Once this restriction changes, I expect we should use generative ones for testing.

**Review by rwhogg — APPROVED** [review]

File: mmv1/third_party/terraform/tests/resource_firestore_field_test.go
Comment (SarahFrench) [reply mmv1/third_party/terraform/tests/resource_firestore_field_test.go]: Sorry, I didn't see this comment 2 days ago. I don't think we're strict about tabs vs spaces, but in this case the varying indentation made it look like the second indexes block was nested inside another block.

**Review by SarahFrench — APPROVED** [review]
LGTM - I'll let the tests run and will merge after assuming all is ok

**Comment by SarahFrench** [comment]
Failed test is unrelated, merging

**Pull request merged**
